keps/sig-network/1111-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,681 @@
+  an HTTP status 503 (StatusServiceUnavailable) is returned, indicating the
+  inverse. Service proxies which implement and support class B type services,
+  must implement the above.
+- For services of class A: it's more of a "free for all". Cloud providers define
+  their own HC, and no one does the same. The following describes what the HC
+  looks like on the three major public cloud providers:
+  - GCP: probes port 10256 (Kube-proxy's healthz port)
+  - AWS: if ELB; probes the first `NodePort` defined on the service spec
+  - Azure: probes all `NodePort` defined on the service spec.
+  
+  All clouds take an approach of trying to ascertain if traffic can be forwarded
+  to the endpoint, which is a completely valid HC for class A services. More is
+  mentioned about this in
+  [Defining what should be "healthy"](#defining-what-should-be-healthy)
+
+To find a solution to the problem of avoiding an ingress outage when a service
+transition from class A <-> B: we will need to align the behavior across service
+classes. That is to say: align service proxies and cloud providers with a
+unified HC which can accurately depict what the load balancer should do for both
+types of services classes, more about this is mentioned in
+["Extend healthCheckNodePort to class A services"](#extend-healthchecknodeport-to-class-a-services)
+
+### Defining what should be "healthy"
+
+When [Defining "healthy"](#defining-healthy), it is important is to keep in mind
+that we are discussing health from the perspective of an
+__application load balancer__.
+
+We could argue that the logic in the KCCM has been flawed all along. Ingress for
+both class A and B services should not care what the state of the node / Kubelet
+is. For class A services the load balancer needs to know if the node it is
+sending traffic to can forward traffic to the node hosting the endpoint. This is
+answered by the services proxy's healthz status (or by targeting the `NodePort`
+of the service). For class B services, the load balancer really needs to know
+where the endpoint is running. We could, and possibly should, extend that
+condition to also validate the state of the service proxy's healthz, since for
+all load balancers the service proxy still needs to write specific rules which
+will forward traffic intended for the node to the endpoint itself. In any case:
+why should the state of the node / Kubelet impact any routing decision? Either
+the proxy is running fine and knows the endpoint is running on the node, or not.
+The only reason to include such logic would be to try to perform pre-mature
+optimization, by trying to predict that if the Kubelet / node is not running
+fine right now, then maybe the endpoint will be scheduled elsewhere soon. But
+that can be very wrong and cut ingress traffic to the application when the node
+experiences short transient errors which will not cause the endpoint to be
+re-scheduled at all.
```

Comment:
This is basically discussing poll vs push, in current implementation the controller receives the data (node status) and act on it ... the service proxy doesn't have an API the controller can check, does it?
the controller polling all the nodes doesn't scale

Reply:
> "should I use this node as data-plane for this service".

is not that the root problem, how to sync pod and node lifecycle?